Connectivity loss & safety

ChrisAxt opened the pull request:
When connection from the app to the broker is lost the car will simulate a "pulling over" maneuver and enters into a parked state

resolves #3
resolves #61 
closes #3 
closes #61 

Notes:
Code implemented by: @gusaxtcha with support/advice from @iachella

ChrisAxt committed:
Merge remote-tracking branch 'origin/bug_SlowdownSmoothly' into Connectivity_Loss_&_Safety

ChrisAxt committed:
Connectivity loss and safety #3
##########################################################
Type:
User Story
Tags
#3
Changes
- Implemented connection loss count
- ConnectivityLoss() with temporary "pull over" maneuver (more for testing and working things out)
- Added an "all" check to is clear to check all sensors
Notes:
- Looking into how to find out if we have lost connection with the app and not just the server, the mqtt.connect() seems to only check the connection to the server, which from the tests done, seems to always be connected.
- Work on this user Story's code had been put on hold until the above is dealt with

ChrisAxt committed:
Connectivity loss and safety #3
##########################################################
Type:
User Story
Tags
#3
Changes
- Implemented last will message (application side)
- Subscribed to connectionLost(last will) topic (Arduino side)
Notes
- Code hasn't been tested yet, will complete and test before a PR.

ChrisAxt committed:
Merge branch 'master' into Connectivity_Loss_&_Safety

ChrisAxt committed:
Connectivity loss and safety #3
##########################################################
Type:
User Story
Tags
#3
Changes
- Fixed problem with imported MqttConnectOptions
Notes

ChrisAxt committed:
Connectivity loss and safety #3
##########################################################
Type:
User Story
Tags
resolves #3
Changes
- Added and improved "pulling over" maneuver within connectivityLoss()
- Added boolean parameter to emergencyBrake() (decides whether to use slowDownSmoothly() or not)
- Removed "all" option from isClear()
- Removed last will code from Dashboard.java
- Added last will code to MqttClient.java
Notes
- All acceptance criteria met, to be tested by terminating the app on android studio (using the stop function), this simulates the ungraceful connection loss.

iachella commented on arduino/smartcar/smartcar.ino:
```diff
@@ -75,9 +78,9 @@ SmartCar car(arduinoRuntime, control, gyroscope, leftOdometer, rightOdometer);
-    connectHost(true); //choosing to connect to localhost.
+    //choose to connect to localhost or external
+    //choose true to connect to localhost.
+    connectHost(false);
```
Should we keep connectHost(true); as default?

ChrisAxt replied:
Depends what you want to connect to, I was testing mostly with the external broker.

alrapal replied:
I would say that the local host make more sense when testing, since there is a lot of "pollution" in the external broker.
But at the end, we just need to switch from one to another :)

iachella commented on UI/build.gradle:
```diff
@@ -5,7 +5,7 @@ buildscript {
     }
     dependencies {
-        classpath "com.android.tools.build:gradle:4.1.1"
+        classpath 'com.android.tools.build:gradle:4.2.0'
```
For this and the next file which has changes about the version used, do we have to manually check and update these versions?

ChrisAxt replied:
These were updated automatically by the IDE.

alrapal replied:
This is the definition of the project, so if it is updated here, when you build this code on your IDE, it will use the version indicated in this file =) (Don't know if that helps understanding haha)

iachella commented on UI/app/src/main/java/com/example/pathfinder/Client/MqttClient.java:
```diff
@@ -23,6 +23,15 @@ public void connect(String username, String password, IMqttActionListener connec
+         * will_Retained - message will be retained bt the server, even after connection loss/ reconnection
+         */
+        String WILL_TOPIC = "/smartcar/connectionLost";
+        String PAYLOAD = "Connection unexpectedly lost";
```
The will_Retained in this case is the String "PAYLOAD" at line 32? If so, is there a reason why it's called PAYLOAD? Just curious, maybe I don't get it in this context

ChrisAxt replied:
I've changed the documentation to be easier to understand, will push new changes soon.

iachella commented on arduino/smartcar/smartcar.ino:
```diff
@@ -102,19 +104,29 @@ void loop()
 void MQTTMessageInput(){
-    if (mqtt.connect("arduino", "public", "public")) {
+
+    if (mqtt.connect("arduino", "public", "public") && mqtt.connected()) {
```
mqtt.connect and mqtt.connected, is the difference between the two that the first is only checking for the specific connection that's in the parameters?

ChrisAxt replied:
This change was redundant, the first is connecting and the second is checking if connected, if one is true, they both are. I have change that and will be included in the latest push.

iachella reviewed (commented):
The code looks good! Just a few questions which I will include as comments :)

ChrisAxt committed:
Connectivity loss and safety #3
##########################################################
Type:
User Story
Tags
fixes #3
Changes
- Improved documentation.
- Deleted redundant check in if statement for mqtt connection.
- small bug fix related to connectionLoss() (If obstacle on the right, car didn't react).
Notes
- All acceptance criteria met, to be tested by terminating the app on android studio (using the stop function), this simulates the ungraceful connection loss.

alrapal reviewed (approved):
Good job with the cleaning and the documentation of the code ! Easier to have a quick understanding of what things are doing. 
Gimmy asked most of my questions, so I have nothing to add to this =)

(I see there are conflicts though, I don't know what does this imply, but it seems that we cannot merge before they are solved, GitHub is clever !)

iachella reviewed (approved):
Checked the changes, thanks for the clarifications!

Edit: Alex, I think the conflicts will be handled with when merging 👍

ediz-genc commented:
The code and your detailed explanations look fine. :) Went through the Q&A parts, I just have two concerns if that is fine to mention/ask! 
(i) How did you test the function? 
(ii) Did you find anything on how to solve the conflicts above? (if you didn't have time to figure that out those conflicts, we can look at it together!)

ChrisAxt committed:
Merge remote-tracking branch 'origin/master' into Connectivity_Loss_&_Safety

# Conflicts:
#	UI/app/src/main/java/com/example/pathfinder/Activities/DashboardActivity.java
#	UI/build.gradle
#	arduino/smartcar/smartcar.ino

ChrisAxt committed:
Connectivity loss and safety #3
##########################################################
Resolved conflicts with master

ChrisAxt commented:
> The code and your detailed explanations look fine. :) Went through the Q&A parts, I just have two concerns if that is fine to mention/ask!
> (i) How did you test the function?
> (ii) Did you find anything on how to solve the conflicts above? (if you didn't have time to figure that out those conflicts, we can look at it together!)

How to test was mentioned in the commit message : 3023b3d however I see now that it's not as visible as I had thought it was. (basically by terminating the app on android studio will simulate a connection loss but any kind of connection loss that does not send a disconnect message will work too.)

The conflicts were dealt with by merging the master into this branch and solving them (Most of the conflicts was due to the documentation I added to some parts of the code that didn't include any.) then with the undated branch there were no conflicts which aloud me to merge and close the request.
